tools/codegen/gen.py

```diff
@@ -78,8 +78,23 @@ def parse_native_yaml(path: str) -> List[NativeFunction]:
         funcs = e.get('func')
         with context(f'in {loc}:\n  {funcs}'):
             rs.append(NativeFunction.from_yaml(e, loc))
+    error_check_native_functions(rs)
     return rs
 
+# Some assertions are already performed during parsing, but those are only within a single NativeFunction.
+# Assertions here are meant to be performed across NativeFunctions.
+def error_check_native_functions(funcs: Sequence[NativeFunction]) -> None:
+    func_map: Dict[OperatorName, NativeFunction] = {}
+    for f in funcs:
+        func_map[f.func.name] = f
+    for f in funcs:
+        if f.structured_delegate is not None:
+            delegate_func = func_map[f.structured_delegate]
+            assert delegate_func.structured, \
+                f"{f.func.name} is marked as a structured_delegate pointing to " \
+                f"{f.structured_delegate}, but {f.structured_delegate} is not marked as structured. " \
+                f"Consider adding 'structured=True' to the delegated operator"
```

Comment:
I think you can do the check instead at StructuredNativeFunctions.from_dict and then you don't need to construct the map here

Reply:
`StructuredNativeFunctions.from_dict` is only aware of the NativeFunctions within a single NativeFunction group. So if I add it in there, we wouldn't be able to enforce the check across NativeFunctionGroup boundaries. That doesn't really fit with the fact that we allow users to pass in any operator name to `structured_delegate`, right?

cc @bhosmer

Reply:
One question is whether we really want to allow users to pass in any operator name. Do we have use cases that need that permissiveness? Anyway though given that's what the system currently allows, this seems like the right check.

Reply:
Yeah, I was wondering that too- I remember us talking a little about it a little during in a composability sync about difference options for the structured syntax in `native_functions.yaml` and this was settled on. If we ever want to change that, e.g. force `structured_delegate` to always delegate to the `out` variant in the same NativeFunctionGroup, it should be pretty easy to do that.